4 benchmark suite setup

tyi1025 opened the pull request:
@burgholzer Hi, I added some stuff. Feel free to look and point out anything that's not good. Besides, I have a couple of questions:
1. Is it possible to run the CI pipeline without pushing? It might be ideal to see whether the CI passes before everything is already on GitHub
2. About the current architecture: should I still implement the VerificationTask.cpp and SimulationTask.cpp or should I just initialize concrete tasks using their respective headers?
3. I added `nlohmann/json`, and now I want to add QuantumCircuit from `extern/ddsim/extern/qfr/mqt/qfr/qiskit/QuantumCircuit.hpp` for the Task classes. Should I then add the whole mqt module?
4. In what context do you mean by "adaptor functions"?
5. I might need some more guidance on the json test setup, but for this part let's wait until I get the source setup right first.

tyi1025 committed:
Merge branch 'main' into 4-benchmark-suite-setup

tyi1025 committed:
✅ add Task, Executor and Result classes

pre-commit-ci[bot] committed:
🎨 pre-commit fixes

burgholzer commented on include/Executor.hpp:
```diff
@@ -0,0 +1,17 @@
+//
+// Created by Tianyi Wang on 5/14/23.
+//
```
Please either use proper file headers (as in the other projects) or remove the auto-generated once as a whole.

burgholzer commented on include/Executor.hpp:
```diff
@@ -0,0 +1,17 @@
+#ifndef DD_EVAL_EXECUTOR_HPP
+#define DD_EVAL_EXECUTOR_HPP
+
+#endif // DD_EVAL_EXECUTOR_HPP
```
These are called header guards. Code should go in between the `#define ...` and the `#endif`. Otherwise they don't serve a purpose at all.
However, it's actually the more modern practice to use `#pragma once` instead.
As for the file here, that would be:
```suggestion
#pragma once
```
Please also adapt the other files accordingly.

burgholzer commented on include/Executor.hpp:
```diff
@@ -0,0 +1,17 @@
+class Executor {
+private:
+  Task* task;
+};
```
As it stands, this class doesn't really serve any purpose. It implements no constructors, has no methods, and can't really be used for anything.

This needs some further work. In the end, this class should define the interface that can be implemented to realise different executors (e.g., the different simulation or verification methods that can process tasks).

As an example: There could be a `SimulationExecutor` and a `VerificationExecutor` that each take a `SimulationTask` and a `VerificationTask`, respectively, execute the respective task and return the corresponding results.

The class hierarchy could look something like
 - `Executor` (the very general interface and everything that is independent of whether it's simulation or verification)
   - `SimulationExecutor` (general interface for simulators; should provide an interface that accepts a `SimulationTask`)
     - `CircuitSimulatorExecutor` (dedicated implementation for one specific simulator that can be benchmarked on various tasks. This class contains all the nitty-gritty details of how to run a specific task with that simulator, includes checks whether a task is executable, and how to collect benchmark metrics.)
     - ... (more different Simulators)
   - `VerificationExecutor` (general interface for verifiers; should provide an interface that accepts a `VerificationTask`)
     - `AlternatingVerificationExecutor` (dedicated implementation for one particular verifier. same as above)
     - ... (more different Verifiers)

Ideally, all of them have some kind of `execute` function that takes a generic `std::shared_ptr<Task> task` and executes the task. I am not yet quite sure where and how to best collect results, but this will quickly become apparent once the first instantiation is implemented.

tyi1025 replied:
Should I include the `CircuitSimulatorExecutor`s and `AlternatingVerificationExecutor`s in this branch as well or should they be done in https://github.com/tyi1025/dd_eval/issues/5 and https://github.com/tyi1025/dd_eval/issues/6 ?

burgholzer replied:
Hm. I'd say one representative for simulation and verification should be included in this PR already to see if everything is working properly.

burgholzer commented on include/Executor.hpp:
```diff
@@ -0,0 +1,17 @@
+
+class Executor {
+private:
+  Task* task;
```
Avoid the use of raw pointers and prefer the use of smart pointers which allow for automatic memory management. (smart pointers are defined in the `<memory>` header)

burgholzer commented on include/Executor.hpp:
```diff
@@ -0,0 +1,17 @@
+  Task* task;
+};
+
+void executeTask(Task* task) { task->execute(); }
```
clang-tidy already warns about this, but just to make it clear: Functions should not be defined in header files in general. That's what source (`cpp`) files are for. 
One of the only exceptions are small functions inside a class that are ok to define directly in the header.

burgholzer commented on include/Task.hpp:
```diff
@@ -0,0 +1,15 @@
+
+class Task {
+public:
+  virtual Result execute() = 0;
```
you need to define a (virtual) destructor. See the corresponding clang-tidy check: https://releases.llvm.org/14.0.0/tools/clang/tools/extra/docs/clang-tidy/checks/cppcoreguidelines-virtual-class-destructor.html

burgholzer replied:
We'll have to see if that signature works as this PR develops. There could be a problem with object slicing due to the return type here being `Result`. If you return an object of a class derived from `Result` this will most likely not work.
If I had a guess, a signature like
```c++
virtual json execute() = 0;
```
could work best, that more or less gets rid of the Results class.

burgholzer commented on include/VerificationTask.hpp:
```diff
@@ -0,0 +1,18 @@
+  //     VerificationTask(QuantumCircuit qc1, QuantumCircuit qc2);
+  // private:
+  //     QuantumCircuit qc1_;
+  //     QuantumCircuit qc2_;
```
Try not to include commented out code (that is not working). No need to commit such changes. You can just keep them locally.

burgholzer commented on src/CMakeLists.txt:
```diff
@@ -6,26 +6,34 @@ add_subdirectory("${PROJECT_SOURCE_DIR}/extern/ddsim" "extern/ddsim"
-# checker/dd/DDSimulationChecker.cpp checker/dd/DDAlternatingChecker.cpp
-# checker/dd/applicationscheme/GateCostApplicationScheme.cpp
-# checker/dd/simulation/StateGenerator.cpp checker/zx/ZXChecker.cpp)
+if(NOT TARGET nlohmann_json)
+  add_subdirectory("${PROJECT_SOURCE_DIR}/extern/ddsim/extern/qfr/extern/json"
+                   "qfr/json" EXCLUDE_FROM_ALL)
+endif()
```
No need to add the `nlohmann_json` library here. It is automatically linked against the QFR target and hence available for all projects that build on the QFR.
```suggestion
```

tyi1025 replied:
~~Thanks for the suggestion. However, my CLion does not seem to recognize the nlohmann_json after I remove this. Should I maybe change some of the CLion settings?~~
Fixed after [this](https://github.com/tyi1025/dd_eval/pull/9/files/f13c774f3d289992408ee0c18f484d521c87740f#r1194430054)

burgholzer commented on src/CMakeLists.txt:
```diff
@@ -6,26 +6,34 @@ add_subdirectory("${PROJECT_SOURCE_DIR}/extern/ddsim" "extern/ddsim"
 
-# add MQT alias
-add_library(MQT::${PROJECT_NAME} ALIAS ${PROJECT_NAME})
+  target_link_libraries(${PROJECT_NAME} PUBLIC nlohmann_json)
```
```suggestion
```
As stated above, the library is directly available. See https://github.com/cda-tum/qfr/blob/a9554ecf51786ce26dba91069607fea2aa2c4937/src/CMakeLists.txt#L76

burgholzer commented on include/SimulationTask.hpp:
Apply the same reasoning here ad for the VerificationTask

burgholzer commented on src/Result.cpp:
```diff
@@ -0,0 +1,17 @@
+// Created by Tianyi Wang on 5/14/23.
+//
+
+#include "../include/Result.hpp"
```
This is not how to include project headers. If everything is set up properly, then
```suggestion
#include "Result.hpp"
```
is enough and the way to go.

tyi1025 replied:
Thanks for clearing this up, but my CLion gives error messages (although they probably don't affect the build) after I remove this, although I included all files into CMake. Do you think I missed something here?

burgholzer replied:
If everything is set up properly, this should not give any error messages. I'll check again.

burgholzer commented on include/Result.hpp:
```diff
@@ -0,0 +1,27 @@
+  std::string toString() const override;
+
+private:
+  std::string name_;
+  double      value_;
```
As pointed out by the clang-tidy check, these do not follow the naming conventions. (do not use trailing underscores for member variables)

burgholzer commented on include/ResultBase.hpp:
```diff
@@ -0,0 +1,20 @@
+  virtual ~ResultBase() = default;
+
+  virtual json        toJson() const   = 0;
+  virtual std::string toString() const = 0;
```
```suggestion
  virtual std::string toString() const { return toJson().dump(2); }
```
No need to make this function abstract.

burgholzer commented on include/ResultBase.hpp:
This should be the class that specifies the bare minimum of statistics that should be provided as a result.
At the very least, this should be some kind of identifier/name and the runtime.
It would be great if this would also include memory consumption (see https://stackoverflow.com/questions/63166/how-to-determine-cpu-and-memory-consumption-from-inside-a-process for pointers on how to obtain this on every platform. Might require some further googling)

As such, this could probably be the class called `Results` and any class that provides more information can derive from this one, e.g., a DDResults class.

tyi1025 replied:
Should the identifier/name be a hash or input by the user/caller?

burgholzer replied:
Good question. I believe the simplest thing would be to have the `Executor` and the `Task` class each provide a `getIdentifier` method and then use those functions to build a name for the result. Or, alternatively just add two fields to the `Result` class: one for the executor and one for the task.

tyi1025 replied:
Should I use macros from C++ to detect the operating system and use different implementations for them?

burgholzer replied:
I think the only OS difference is gathering memory statistics. Everything else should be platform independent.
I'd advise creating a separate file that encapsulates the OS-dependent behavior.

You can, and probably have to use macros for that.

burgholzer replied:
Some good references:
- https://github.com/SaulBerrenson/sys_info
- https://learn.microsoft.com/en-us/windows/win32/psapi/collecting-memory-usage-information-for-a-process
- https://stackoverflow.com/questions/63166/how-to-determine-cpu-and-memory-consumption-from-inside-a-process

burgholzer commented on include/Result.hpp:
This class doesn't really make sense at the moment, as it does not really implement anything useful as a "Result". See also the comments in the base class. I'd suggest removing it for now.

burgholzer commented on include/Task.hpp:
```diff
@@ -0,0 +1,15 @@
+
+class Task {
+public:
+  virtual Result execute() = 0;
```
I would not have expected the `Task` to have an `execute` function. That's the job of the `Executor`.
The `Task` just describes an instance to be benchmarked.
For Simulation tasks that would be a single circuit, for verification tasks it would be two circuits.
The `Task` class then only needs appropriate constructors and helper functions.

tyi1025 replied:
My initial idea was to apply a Command design pattern and have a void execute() function from the invoker (executor) which calls the execute() function in the command (task). This approach although kind of confused me when I was looking at the benchmarking stuff in qmap and thinking about how to implement it. I'll give it some more thought👍

burgholzer reviewed (changes requested):
Just had a quick glance through everything here.
I added quite some rather general comments regarding coding best practices and such.
The other half of the comments is topic related. As of now, there is not much here, which makes it kind of hard to judge.
The goal for this PR should be to have a setup that allows to specify Simulation and Verifikation Executors and Tasks. In addition, this should also be tested.

Note, that this only has to be 
 - a single, basic simulator 
 - a single verification method 
 - only the most basic metrics (some general metrics as described in the comments and one special DD metric)
 - just simple test circuits. One for simulation and two for verification.

That should suffice to find out whether the general framework works. Based on that:
 - Further Simulation Executors will come in #5 
 - Further Verification Executors will come in #6 
 - More benchmarks will be added as part of #5 and #6
 - More evaluation metrics will be added as part of #7

burgholzer commented:
> @burgholzer Hi, I added some stuff. Feel free to look and point out anything that's not good. Besides, I have a couple of questions:
> 
> 1. Is it possible to run the CI pipeline without pushing? It might be ideal to see whether the CI passes before everything is already on GitHub

Almost everything can be run before pushing:
- You can build and run tests locally before pushing
- `clang-format` and `clang-tidy` are natively integrated into CLion and should per default show warnings and reformat files properly
- you can simply run `pre-commit` locally via `pre-commit run -a` or even easier, have it automatically run as a git hook by installing them once with `pre-commit install`
- coverage can also be collected directly in CLion via the respective run option

That leaves only builds on other platforms than your host and the GitHub CodeQL checks which are fine to just run in CI.

> 2. About the current architecture: should I still implement the VerificationTask.cpp and SimulationTask.cpp or should I just initialize concrete tasks using their respective headers?

I hope this should be answered by the Review comments. Tasks should really only describe ... well... tasks to be passed to a simulator or a verifier.

> 3. I added `nlohmann/json`, and now I want to add QuantumCircuit from `extern/ddsim/extern/qfr/mqt/qfr/qiskit/QuantumCircuit.hpp` for the Task classes. Should I then add the whole mqt module?

Also see the comments on `nlohmann_json`, which is available by default. Note that you are looking for the wrong quantum circuit class here. The QFR has the `QuantumComputation` class, which lives in the `qc` namespace and can be directly imported via `#include "QuantumComputation"`.

> 4. In what context do you mean by "adaptor functions"?

In the sense of the `Executor` classes. They "adapt" the various individual simulators (etc.) to conform to a single, standardised interface and make sure to get the proper results from the execution. 

> 5. I might need some more guidance on the json test setup, but for this part let's wait until I get the source setup right first.

You should be able to (more-or-less) copy and paste this from the QMAP setup I linked.
Note that you should probably create two test targets in the long run: one that is intended to run in CI and runs quickly with the purpose of testing the overall functionality. and one that runs the "real" evaluations that we are interested in and will take quite a while to run. 
That second target should be excluded from CI and things like coverage as it is never gonna be run in the CI environment.

burgholzer commented on src/CMakeLists.txt:
```diff
@@ -6,26 +6,34 @@ add_subdirectory("${PROJECT_SOURCE_DIR}/extern/ddsim" "extern/ddsim"
-# link to the MQT libraries
-target_link_libraries(${PROJECT_NAME} INTERFACE MQT::ddsim MQT::qcec)
+  # link to the MQT libraries
+  target_link_libraries(${PROJECT_NAME} INTERFACE MQT::ddsim MQT::qcec)
```
```suggestion
  target_link_libraries(${PROJECT_NAME} PUBLIC MQT::ddsim MQT::qcec)
```
One important change I forgot. Should also fix the nlohman_json issue.

burgholzer commented on src/CMakeLists.txt:
```diff
@@ -6,26 +6,34 @@ add_subdirectory("${PROJECT_SOURCE_DIR}/extern/ddsim" "extern/ddsim"
-                            ${PROJECT_BINARY_DIR}/include)
+  # set include directories
+  target_include_directories(
+    ${PROJECT_NAME} INTERFACE ${PROJECT_SOURCE_DIR}/include
```
```suggestion
    ${PROJECT_NAME} PUBLIC ${PROJECT_SOURCE_DIR}/include
```
also needs to be public so that the library files itself can access the headers and not only any target that links against the DDEval library.

tyi1025 replied:
Life saved👍

tyi1025 committed:
First review suggestions taken

tyi1025 committed:
Merge remote-tracking branch 'origin/4-benchmark-suite-setup' into 4-benchmark-suite-setup

pre-commit-ci[bot] committed:
🎨 pre-commit fixes

tyi1025 commented:
@burgholzer It seems that CLion still does not recognize the two exemplary .cpp files in my test folder ("this file does not belong to any project target..."), despite them being in the CMake file. After some investigation, I think the problem might be that there is something I didn't configure about building the tests. Should I fix it or is it irrelevant for now?

tyi1025 commented:
> In the sense of the `Executor` classes. They "adapt" the various individual simulators (etc.) to conform to a single, standardised interface and make sure to get the proper results from the execution.
This means that the adapter functions should be specified in individual subclasses of Executor instead of declared in the headers, is it correct?

burgholzer commented:
> @burgholzer It seems that CLion still does not recognize the two exemplary .cpp files in my test folder ("this file does not belong to any project target..."), despite them being in the CMake file. After some investigation, I think the problem might be that there is something I didn't configure about building the tests. Should I fix it or is it irrelevant for now?

Have you configured the current CMake target with `-DBUILD_DD_EVAL_TESTS=ON`?
And have you actually added the test files to a test target in the test/CMakeLists.txt file?
Last thing: maybe try clearing the CMakeCache.
That should be something you get fixed in your setup.

tyi1025 commented:
> > @burgholzer It seems that CLion still does not recognize the two exemplary .cpp files in my test folder ("this file does not belong to any project target..."), despite them being in the CMake file. After some investigation, I think the problem might be that there is something I didn't configure about building the tests. Should I fix it or is it irrelevant for now?
> 
> Have you configured the current CMake target with `-DBUILD_DD_EVAL_TESTS=ON`? And have you actually added the test files to a test target in the test/CMakeLists.txt file? Last thing: maybe try clearing the CMakeCache. That should be something you get fixed in your setup.

Thanks, it's fixed now. The problem was I wrote `...DDEVAL_TESTS...` instead of `...DD_EVAL_TESTS...` in my CLion run configuration🤦‍♂️

tyi1025 commented:
> In the sense of the `Executor` classes. They "adapt" the various individual simulators (etc.) to conform to a single, standardised interface and make sure to get the proper results from the execution.

This means that the adapter functions should be specified in individual subclasses of Executor instead of declared in the headers, is it correct?

burgholzer commented:
> > In the sense of the `Executor` classes. They "adapt" the various individual simulators (etc.) to conform to a single, standardised interface and make sure to get the proper results from the execution.
> 
> 
> 
> This means that the adapter functions should be specified in individual subclasses of Executor instead of declared in the headers, is it correct?

Yeah. I suspect some of these subclasses will be rather small or look similar. But this should allow for some more freedom when it comes to implementing various executors.

tyi1025 committed:
Refactored executor-task architecture

tyi1025 commented:
@burgholzer Do you think the code structure as of now makes sense? Please note that the subclasses of `Executor` and `Task` are not finished yet, which is why I didn't formally re-request a review.

tyi1025 committed:
Refactored naming

burgholzer commented:
> @burgholzer Do you think the code structure as of now makes sense? Please note that the subclasses of `Executor` and `Task` are not finished yet, which is why I didn't formally re-request a review.

It's going into the right direction. It's still hard to judge because there are no tests and it isn't implemented end-to-end.
Please try to get a complete setup done and iterate on that. You have a lot of freedom on how to realize the overall architecture. It's on you to implement something that works and can be used. You shouldn't need my feedback on every single step along the way.

tyi1025 commented:
> > @burgholzer Do you think the code structure as of now makes sense? Please note that the subclasses of `Executor` and `Task` are not finished yet, which is why I didn't formally re-request a review.
> 
> It's going into the right direction. It's still hard to judge because there are no tests and it isn't implemented end-to-end.
> Please try to get a complete setup done and iterate on that. You have a lot of freedom on how to realize the overall architecture. It's on you to implement something that works and can be used. You shouldn't need my feedback on every single step along the way.

Understand, I will keep that in mind.
Although we agreed on a flexible schedule, could you maybe give me a time recommendation, e.g. when should I finish the benchmarking tasks to not fall behind schedule (primarily the TUM project deadline in the week of October 16th), so that I'll have a better grasp on the big picture? I feel that some of my (stupid) questions might come from this "anxiety" because of the lack of knowledge of a "deadline".

burgholzer commented:
I understand that feeling.
However, there is lots of time until October and lots of flexibility when it comes to the extent of this project. So there is really no need to worry. You'll get something done for sure.

As for more precise deadlines: the earlier this benchmark suite is "ready" or at least set up, the earlier you can work on extracting relevant statistics from the DD Package (which will definitely also require some time).
After that the real improvements on the DD Package happen.
If I had a guess then the benchmark suite setup here shouldn't take much more than a week or so. 
Collecting the benchmarks and setting up the various simulators / verifiers might take another 2-3 days.
Gathering/logging all the statistics will probably require 2-3 weeks as you really have to learn about the DD package architecture at that point.
The rest of the time can then be spent on improvements of the package and aggregation/visualization of the benchmark results.

Hope that helps.

tyi1025 committed:
Merge remote-tracking branch 'origin/main' into 4-benchmark-suite-setup

tyi1025 committed:
Refactor simulation executor

tyi1025 committed:
Refactored executor and simulatorexecutor

tyi1025 committed:
Added dummy test (other tests WIP)

tyi1025 committed:
Refactored test code to not let the task consume the qc

tyi1025 committed:
Fixed a getter function in SimulationExecutor to address a warning

tyi1025 committed:
Move non-parameterized simple test case to a separate file

tyi1025 committed:
Clean up the test suite for SimulatorExecutors

tyi1025 committed:
Remove redundant comments

tyi1025 committed:
Replace absolute path with relative path

tyi1025 committed:
Refactor the results json

tyi1025 committed:
Added verification executor classes and a simple test

tyi1025 committed:
Delete temporarily obsolete Results class

tyi1025 committed:
Add qcec tests (incl. refactoring ddsim tests to avoid name collisions)

tyi1025 committed:
Refactored exec and task identifier

tyi1025 committed:
Refactor Task and Executor inheritance

tyi1025 committed:
Clean-up

tyi1025 committed:
Make sure the alternating checker was run in alternating executor

tyi1025 committed:
Make sure the alternating checker was run in alternating executor

tyi1025 committed:
Merge remote-tracking branch 'origin/4-benchmark-suite-setup' into 4-benchmark-suite-setup

tyi1025 committed:
Fix CI issues (WIP)

tyi1025 committed:
Refactor architecture: set all necessary members inside of constructor

tyi1025 committed:
Refactor simulator test to avoid empty circuits

tyi1025 committed:
Refactor verification tests to avoid empty circuits

tyi1025 committed:
Remove obsolete code for coverage

burgholzer commented on include/Task.hpp:
```diff
@@ -0,0 +1,11 @@
+
+#include <nlohmann/json.hpp>
+
+using json = nlohmann::json;
```
This include and using directive are not used here. Consider removing them and only importing json when necessary.
The motto in C++ is IWYU (include what you use).

burgholzer commented on src/VerificationTask.cpp:
Same as for the `SimulationTask`

Methods like these, that just take a single line of code, can just be inlined in the header file.
In this case, you do not need the source (`cpp`) file at all.

burgholzer commented on src/SimulationTask.cpp:
Methods like these, that just take a single line of code, can just be inlined in the header file.
In this case, you do not need the source (`cpp`) file at all.

burgholzer commented on include/SimulationTask.hpp:
You are missing the `<memory>` include here

burgholzer commented on include/VerificationTask.hpp:
You are missing the `<memory>` include here

burgholzer commented on include/Executor.hpp:
```diff
@@ -0,0 +1,17 @@
+  virtual std::string getIdentifier() = 0;
+
+private:
+  std::unique_ptr<Task> task;
```
A unique pointer does not really make sense here from a conceptional point of view.
You do not want the Executor to _own_ the task. A single task should be executable by multiple Executors (without any copying/duplication/moving)

burgholzer commented on include/Executor.hpp:
```diff
@@ -0,0 +1,17 @@
+  virtual ~Executor() = default;
+  explicit Executor() = default;
+
+  virtual json        executeTask()   = 0;
```
I think this interface could be improved. I don't see real value being added by having the executor _own_ the task (see also the comment below).
How about something like
```c++
virtual json execute(const Task& task) = 0;
```
?
That directly communicated that this executes a certain task and yields `JSON` as a result.

Considering that you might want to make sure that a certain Executor only gets certain jobs to process, it might be worth turning the executor class into a template class, where the template parameter is the Task class. Some boilerplate code for that:
```c++
#include <type_traits>

template<class T>
class Executor {
  static_assert(std::is_base_of_v<Task, T>);
public:
  virtual ~Executor() = default;
  explicit Executor() = default;

  virtual json execute(const T& task) = 0;
}
```

Then, any SimulationExecutor just inherits from `Executor<SimulationTask>`, etc.

burgholzer commented on include/Executor.hpp:
```diff
@@ -0,0 +1,17 @@
+
+#include "Task.hpp"
+
+#include <memory>
```
In case of the changes below, this include won't be necessary.

burgholzer commented on include/SimulationExecutor.hpp:
This class doesn't really add any value and can be removed. Especially considering the comments in the `Executor` header.

burgholzer commented on include/VerificationExecutor.hpp:
This class doesn't really add any value and can be removed. Especially considering the comments in the `Executor` header.

burgholzer commented on include/SimulationTask.hpp:
Just a suggestion:
you might want to create a directory `tasks` and move all the specific tasks (such as the file here) to that folder. That should create a cleaner folder structure.

Similarly, you can move all the specific executors to an `executors` directory that can have further subdirectories for `simulation` and `verification`.

tyi1025 replied:
Ok, sounds good. Should Task and Executor stay in the root include directory though?

burgholzer replied:
I'd say yes, but have no strong opinion here.

burgholzer commented on include/Executor.hpp:
You are missing the json include and using directive here.

burgholzer commented on include/TestHelpers.hpp:
if this is a helper for the tests, then it should go into the `test` directory and not into the main library.

burgholzer commented on src/TestHelpers.cpp:
```diff
@@ -0,0 +1,11 @@
+    const auto& key   = it.key();
+    const auto& value = it.value();
+    std::cout << key << ": " << value << std::endl;
+  }
```
this function isn't necessary at all. You can just call `std::cout << j.dump(2U) << std::endl;` to get a formatted JSON output at any time.

burgholzer commented on src/SimulationExecutor.cpp:
Per comments in other files, this is not necessary and if it were, it should be defined in the header file.

burgholzer commented on src/VerificationExecutor.cpp:
Per comments in other files, this is not necessary and if it were, it should be defined in the header file.

burgholzer commented on src/CircuitSimulatorExecutor.cpp:
```diff
@@ -0,0 +1,38 @@
+
+json CircuitSimulatorExecutor::executeTask() {
+  json result;
+  auto start = std::chrono::high_resolution_clock::now();
```
Prefer to use https://en.cppreference.com/w/cpp/chrono/steady_clock here. See e.g. https://youtrack.jetbrains.com/issue/KT-33934

burgholzer commented on src/CircuitSimulatorExecutor.cpp:
```diff
@@ -0,0 +1,38 @@
+  for (const auto& entry : m) {
+    j[entry.first] = entry.second;
+  }
+}
```
This is not necessary.
You can just use 
```c++
json j = <whatever map you have>
```

burgholzer commented on src/CircuitSimulatorExecutor.cpp:
```diff
@@ -0,0 +1,38 @@
+}
+
+std::string CircuitSimulatorExecutor::getIdentifier() {
+  return "circ_sim_exe" + mTask->getIdentifier();
```
If you implement the other changes, then it will become clear, that this kind of identifier here does not really make sense.
The identifier of the executor should only include the name of the executor itself and not the task.
Here, you could just use `circuit_simulator` for example.

(also, consider moving that function to the header file)

burgholzer commented on src/CircuitSimulatorExecutor.cpp:
```diff
@@ -0,0 +1,38 @@
+  auto start = std::chrono::high_resolution_clock::now();
+
+  const auto results = mCircuitSimulator->simulate(1024U);
+  toJson(result, results);
```
```suggestion
  result["measurement_results"] = mCircuitSimulator->simulate(1024U);
```
should technically work and directly create some json out of the measurement results map.

burgholzer commented on src/CircuitSimulatorExecutor.cpp:
```diff
@@ -0,0 +1,38 @@
+
+  const auto results = mCircuitSimulator->simulate(1024U);
+  toJson(result, results);
+  // Add memory usage
```
This function should also add the respective identifiers to the result, i.e.,
```c++
result["executor"] = getIdentifier();
result["task"] = task.getIdentifier();
```

burgholzer commented on src/CircuitSimulatorExecutor.cpp:
```diff
@@ -0,0 +1,38 @@
+
+  auto stop = std::chrono::high_resolution_clock::now();
+  auto runtime =
+      std::chrono::duration_cast<std::chrono::microseconds>(stop - start);
```
both of these can be `const`

burgholzer commented on src/CircuitSimulatorExecutor.cpp:
```diff
@@ -0,0 +1,38 @@
+  result["runtime"] = runtime.count();
+  std::string const identifier =
+      this->getTask()->getIdentifier() + "_" + this->getIdentifier();
+  result["identifier"] = identifier;
```
Per the comment above this is redundant and can be removed.

burgholzer commented on src/CircuitSimulatorExecutor.cpp:
```diff
@@ -0,0 +1,38 @@
+    std::unique_ptr<SimulationTask> simulationTask) {
+  mTask   = std::move(simulationTask);
+  auto qc = std::make_unique<qc::QuantumComputation>(mTask->getQc()->clone());
+  mCircuitSimulator = std::make_unique<CircuitSimulator<>>(std::move(qc));
```
I'd suggest not allocating the CircuitSimulator as a member of the executor, but to create a new instance for every task. As it is right now, the class only handles a single task anyway, but this is subject to change after following the other comments.
I would suggest to remove this member here and put the respective code into the `execute` function.
Specifically, clone the circuit and create the circuit simulator there.

Side note: one of the main benefits of this is that every task is ensured to get a "fresh" DD package instance. So the performance is independent of previously executed tasks.

tyi1025 replied:
After I move these into the execute function, where should the start of the clock be? Should it be after these object initializations (hence only recording the time of execution) or should it include everything?

burgholzer replied:
Good question.
Why not track both? That is, construction time and execution time?
For example, I would expect the construction time to get lower once the package data structures can be dynamically resized since a smaller allocation will be required in the beginning.

burgholzer commented on include/AlternatingVerificationExecutor.hpp:
Essentially everything I wrote for the Circuit Simulator Executor also holds here. You should be able to adapt that accordingly.

burgholzer commented on src/AlternatingVerificationExecutor.cpp:
```diff
@@ -0,0 +1,36 @@
+    std::unique_ptr<VerificationTask> verificationTask) {
+  mTask    = std::move(verificationTask);
+  auto qc1 = mTask->getQc1()->clone();
+  auto qc2 = mTask->getQc2()->clone();
```
No need to clone here. The equivalence checking manager itself takes care of cloning the circuits.
(QCEC makes live a little easier here as compared to DDSIM)

you should be able to just pass `*qc1` and `*qc2` to the constructor.

burgholzer commented on src/AlternatingVerificationExecutor.cpp:
```diff
@@ -0,0 +1,36 @@
+  mEquivalenceCheckingManager->setAlternatingChecker(true);
+  mEquivalenceCheckingManager->setSimulationChecker(false);
+  mEquivalenceCheckingManager->setConstructionChecker(false);
+  mEquivalenceCheckingManager->setZXChecker(false);
```
Just for ensuring that this is future-proof: The ECM has the following method for disabling all checks: https://github.com/cda-tum/qcec/blob/a1056e0c5c158a4a3b07899cd516f9c190549635/include/EquivalenceCheckingManager.hpp#L120
Consider calling that and then only enabling the checker you want.

burgholzer commented on src/AlternatingVerificationExecutor.cpp:
```diff
@@ -0,0 +1,36 @@
+  auto start = std::chrono::high_resolution_clock::now();
+
+  mEquivalenceCheckingManager->run();
+  result = mEquivalenceCheckingManager->getResults().json();
```
Similar to the suggestion in the Simulation Executor: consider moving the results of the equivalence check into a separate JSON entry to not confuse them with the benchmarking results.
Something like `check_results` could be an appropriate name.

burgholzer commented on test/circuits.json:
```diff
@@ -0,0 +1,7 @@
+  {
+    "description": "two_qubit_circuit_with_two_x_gates",
+    "initial_circuit": "OPENQASM 2.0;include \"qelib1.inc\";qreg q[2];x q[0];x q[1];\n",
+    "expected_11": 1024
```
how do you plan to generalize that concept of expected results to a whole range of benchmarks across various numbers of qubits and algorithms?
Consider revising this in some fashion that is a little more generalizable.
One possible suggestion would be to specify parts of how the `result` json should look like and assert that in the tests. Might not be the best suggestion though.

burgholzer commented on test/equi_circuits.json:
```diff
@@ -0,0 +1,8 @@
+    "description": "simple_circuits",
+    "initial_circuit1": "OPENQASM 2.0;include \"qelib1.inc\";qreg q[2];x q[0];x q[1];\n",
+    "initial_circuit2": "OPENQASM 2.0;include \"qelib1.inc\";qreg q[2];x q[0];x q[1];x q[0];x q[0];\n",
+    "expected_equivalence": "equivalent"
```
Also here it might be better to be a little more flexible, e.g., by specifying part of the expected `result` JSON.

burgholzer commented on test/circuits.json:
Including more than one task here probably would be good to check that everything is working as intended. Same goes for verification.

burgholzer commented on test/test_qcecexec_simple.cpp:
can be removed as it doesn't add much value and is completely covered by the other tests

burgholzer commented on test/test_simexec_simple.cpp:
can be removed as it doesn't add much value

burgholzer commented on test/test_simexec.cpp:
```diff
@@ -0,0 +1,65 @@
+  void SetUp() override {
+    test = GetParam();
+
+    std::stringstream      ss(test.initialCircuit);
+    qc::QuantumComputation qc{};
+    qc.import(ss, qc::Format::OpenQASM);
+    std::cout << "Circuit:\n" << qc;
+
+    qcSim          = std::make_unique<qc::QuantumComputation>(qc.clone());
+    simulationTask = std::make_unique<SimulationTask>(std::move(qcSim));
```
This kind of construction does not look efficient and/or convenient at all.
```suggestion
    std::stringstream      ss(test.initialCircuit);
    auto qc = std::make_unique<qc::QuantumComputation>():
    qc->import(ss, qc::Format::OpenQASM);
    std::cout << "Circuit:\n" << qc;

    simulationTask = std::make_unique<SimulationTask>(std::move(qc));
```
completely eliminates that one copy.
Furthermore, the simulation task doesn't need to be a unique_ptr. It can just be a regular, plain old `SimulationTask`

burgholzer reviewed (changes requested):
Thanks for your work on this. Seems like you are getting more and more familiar with everything. I still believe this needs some major changes before actually being useful. You'll find many suggestions for changes in the comments below. I hope the respective explanations and suggestions are clear enough so that you can work with them.
I have the feeling that this PR should be more or less good to go after all the issues have been addressed.
Most of the feedback is rather straight to the point and should be easy to implement. The most open question probably concerns the tests and how to check the expected behavior. This might involve some creativity from your side.

tyi1025 committed:
Apply suggestions from code review

Co-authored-by: Lukas Burgholzer <lukas.burgholzer@jku.at>

pre-commit-ci[bot] committed:
🎨 pre-commit fixes

tyi1025 committed:
Refactoring the source code according to the review suggestions

tyi1025 committed:
Refactor test files to remove redundant copy

tyi1025 committed:
Track both times in execution

tyi1025 committed:
Fix ci issues

tyi1025 committed:
Fix CI issues (continued)

tyi1025 committed:
Refactor tests to be more flexible

burgholzer committed:
🎨 assorted cleanups

burgholzer committed:
⚗️ try newer cpp-linter config

burgholzer committed:
🩹 fix CI coverage configuration

burgholzer committed:
🚨 another clang-tidy warning fixed

burgholzer committed:
♻️ remove expected output from simulation test

the criterion is just not generalizable in a reasonable way

burgholzer committed:
♻️ some restructuring

Also make use of forward declarartions to reduce the number of imports in headers

burgholzer reviewed (approved):
Should be good to go 👍🏼
